detecting_metamask.md

```diff
@@ -2,65 +2,64 @@
 
 MetaMask enables your website to read from the blockchain and propose transactions to the current user. To enjoy these benefits, you have to first check if the user has an Ethereum-compatible browser like MetaMask.
 
-## Detecting MetaMask
-
-Metamask/Mist currently inject their interface into page as global `web3` object.
+## What is an Ethereum provider?
 
-```javascript
-window.addEventListener('load', function() {
+A provider is a low-level object with one primary method: `sendAsync(options, callback)`. The `options` object has several useful fields: `method` (the method name), `from` (the sender address), `params` (an array of parameters to send to that method).
 
-  // Checking if Web3 has been injected by the browser (Mist/MetaMask)
-  if (typeof web3 !== 'undefined') {
+These methods all correspond directly to the options in [the RPC provider spec](https://github.com/ethereum/wiki/wiki/JSON-RPC).
 
-    // Use the browser's ethereum provider
-    var provider = web3.currentProvider
+## Detecting MetaMask
 
-  } else {
-    console.log('No web3? You should consider trying MetaMask!')
-  }
+MetaMask and other privacy-focused browsers require dapps to request a provider before using it. Some older browsers may still inject a Web3 instance automatically, so it's best practice to handle both cases:
 
-})
+```javascript
+// If web3 is not injected (modern browsers)...
+if (typeof web3 === 'undefined') {
+    // Listen for provider injection
+    window.addEventListener('message', ({ data }) => {
+        if (data && data.type && data.type === 'ETHEREUM_PROVIDER_SUCCESS') {
+            // Use injected provider, start dapp...
+            web3js = new Web3(ethereum);
+        }
+    });
+    // Request provider
+    window.postMessage({ type: 'ETHEREUM_PROVIDER_REQUEST' }, '*');
+}
+// If web3 is injected (legacy browsers)...
+else {
+    // Use injected provider, start dapp
+    web3js = new Web3(web3.currentProvider);
+}
 ```
 
-The `provider` object is a low-level object with just one supported method: `provider.sendAsync(options, callback)`.
-
-The `options` object has several useful fields: `method` (the method name), `from` (the sender address), `params` (an array of parameters to send to that method).
-
-These methods all correspond directly to the options in [the RPC provider spec](https://github.com/ethereum/wiki/wiki/JSON-RPC).
-
-To see if the injected provider is from MetaMask, you can check `web3.currentProvider.isMetaMask`.
+After user approval, a provider object will be injected as a `window.ethereum` global variable. To see if the injected provider is from MetaMask, you can check `ethereum.isMetaMask`.
```

Comment:
Is this going to follow the provider 2.0 spec? I know we asked that before, but it sure is a good opportunity to "break it once and let it heal" instead of breaking it twice. I have no idea how practical that would be, and it likely is not practical, in which case ignore it.